File: word-pattern/c++/word-pattern.cpp
#include <iostream>
#include <map>
#include <set>
#include <vector>

using namespace std;

vector<string> splitString(string inputString, string delimiter) {
    size_t pos = 0;
    vector<string> results;
    while ((pos = inputString.find(delimiter)) != string::npos) {
        string word = inputString.substr(0, pos);
        inputString = inputString.substr(pos + 1);
        results.push_back(word);
    }
    results.push_back(inputString);
    return results;
}

bool wordPattern(string pattern, string s) {
    map<char, string> patternMap;
    vector<string> words = splitString(s, " ");

    if (words.size() != pattern.length()) return false;

    set<string> visitedWords;
    for (int i = 0; i < words.size(); i++) {
        if (patternMap.count(pattern[i])) {
            if (patternMap[pattern[i]] != words[i])
                return false;
            else
                continue;
        }

        if (visitedWords.find(words[i]) != visitedWords.end()) return false;

        patternMap[pattern[i]] = words[i];
        visitedWords.insert(words[i]);
    }

    return true;
}

int main() {
    string testPattern = "abba";
    string testWords = "dog cat cat dog";

    bool testResult = wordPattern(testPattern, testWords);
    
    if(testResult == true)
        cout << "Pattern : " << testPattern << " matches with: " << testWords;
    else
        cout << "Pattern : " << testPattern << " doesn't match with: " << testWords;

    return 0;
}

File: word-pattern/python3/word-pattern.py
def wordPattern(pattern: str, s: str) -> bool:
    map_chars = {}
    map_words = {}
        
    words = s.split(' ')
        
    if len(words) != len(pattern):
        return False
        
    for char, word in zip(pattern, words):
        if char not in map_chars:
            if word not in map_words:
                map_chars[char] = word
                map_words[word] = char
            else:
                if map_words[word] != char:
                    return False
        else:
            if map_chars[char] != word:
                return False
    return True

test_pattern = "abba"
test_words   = "dog cat cat dog"

if wordPattern(test_pattern, test_words) == True:
    print(f"Pattern {test_pattern} matches with words {test_words}!")
else:
    print(f"Pattern {test_pattern} doesn't match with words {test_words}!")